File: cmd/ambience/web/effects/beach.js
'use strict';
(function (api) {
	const { makeRNG, jitterInt, clamp01, hslToRGB, positiveMod } = api._helpers;

	const DEFAULTS = {
		intro_dur: 55,
		intro_tide: 0.18,
		ending_dur: 65,
		ending_linger: 18,
		ending_wet: 0.1,
		shoreline: 0.58,
		tide_amp: 6,
		wave_amp: 2.4,
		wave_freq: 0.18,
		speed: 0.1,
		slope: 0.16,
		foam: 0.36,
		shimmer: 0.22,
		hue: 198,
		hue_sp: 16,
		sat: 0.5,
		lmin: 0.28,
		lmax: 0.82,
		high_tide_p: 0,
		low_tide_p: 0,
		foam_burst_p: 0,
		high_tide_dur: 60,
		high_tide_push: 1.4,
		low_tide_dur: 58,
		low_tide_pull: 1.2,
		foam_burst_dur: 34,
		foam_burst_mult: 1.9,
	};

	function applyDefaults(cfg) {
		const c = Object.assign({}, DEFAULTS, cfg || {});
		if (c.intro_dur <= 0) c.intro_dur = DEFAULTS.intro_dur;
		c.intro_tide = clamp01(c.intro_tide);
		if (c.ending_dur <= 0) c.ending_dur = DEFAULTS.ending_dur;
		if (c.ending_linger < 0) c.ending_linger = 0;
		c.ending_wet = clamp01(c.ending_wet);
		if (c.shoreline <= 0) c.shoreline = DEFAULTS.shoreline;
		if (c.tide_amp <= 0) c.tide_amp = DEFAULTS.tide_amp;
		if (c.wave_amp <= 0) c.wave_amp = DEFAULTS.wave_amp;
		if (c.wave_freq <= 0) c.wave_freq = DEFAULTS.wave_freq;
		if (c.speed <= 0) c.speed = DEFAULTS.speed;
		if (c.foam <= 0) c.foam = DEFAULTS.foam;
		if (c.shimmer <= 0) c.shimmer = DEFAULTS.shimmer;
		if (c.hue === 0) c.hue = DEFAULTS.hue;
		if (c.hue_sp < 0) c.hue_sp = 0;
		if (c.sat <= 0) c.sat = DEFAULTS.sat;
		if (c.lmin <= 0) c.lmin = DEFAULTS.lmin;
		if (c.lmax <= 0) c.lmax = DEFAULTS.lmax;
		if (c.lmax < c.lmin) [c.lmin, c.lmax] = [c.lmax, c.lmin];
		if (c.high_tide_dur <= 0) c.high_tide_dur = DEFAULTS.high_tide_dur;
		if (c.high_tide_push <= 0) c.high_tide_push = DEFAULTS.high_tide_push;
		if (c.low_tide_dur <= 0) c.low_tide_dur = DEFAULTS.low_tide_dur;
		if (c.low_tide_pull <= 0) c.low_tide_pull = DEFAULTS.low_tide_pull;
		if (c.foam_burst_dur <= 0) c.foam_burst_dur = DEFAULTS.foam_burst_dur;
		if (c.foam_burst_mult <= 0) c.foam_burst_mult = DEFAULTS.foam_burst_mult;
		return c;
	}

	class Beach {
		constructor(w, h, cfg, seed) {
			this.w = w;
			this.h = h;
			this.grid = new Uint8ClampedArray(w * h * 3);
			this.seed = Number(seed || Date.now());
			this.tick = 0;
			this.timers = {};
			this.values = {};
			this.cfg = applyDefaults(cfg);
		}

		setConfig(cfg) {
			this.cfg = applyDefaults(Object.assign({}, this.cfg, cfg));
		}

		restoreSnapshot(snap) {
			const state = snap.state || snap;
			this.setConfig(snap.config || {});
			this.tick = state.tick || snap.tick || 0;
			this.timers = Object.assign({}, state.timers || {});
			this.values = Object.assign({}, state.values || {});
			if (typeof snap.seed === 'number') this.seed = snap.seed;
			if (snap.gridW > 0 && snap.gridH > 0) {
				this.w = snap.gridW;
				this.h = snap.gridH;
			}
		}

		_eventRng(salt) {
			return makeRNG(((this.seed >>> 0) ^ (((this.tick + salt) * 2654435761) >>> 0)) >>> 0);
		}

		_hash(index) {
			const x = Math.sin((this.seed * 0.000001 + index * 12.9898) * 43758.5453);
			return x - Math.floor(x);
		}

		_phaseProgress(total, left) {
			if (left <= 1 || total <= 1) return 1;
			const elapsed = total - left;
			if (elapsed <= 0) return 0;
			return clamp01(elapsed / Math.max(1, total - 1));
		}

		_fillCell(ctx, sx, sy, ceilSx, ceilSy, x, y, w, h, color, alpha) {
			ctx.fillStyle = color;
			ctx.globalAlpha = alpha == null ? 1 : alpha;
			ctx.fillRect(Math.floor(x * sx), Math.floor(y * sy), Math.max(1, Math.ceil(w * sx || ceilSx)), Math.max(1, Math.ceil(h * sy || ceilSy)));
			ctx.globalAlpha = 1;
		}

		triggerEvent(name) {
			const rng = this._eventRng(name.length + 61);
			switch (name) {
				case 'high-tide':
					this.timers['high-tide'] = jitterInt(rng, this.cfg.high_tide_dur, 0.3);
					this.timers['low-tide'] = 0;
					this.values.tide_bias = this.cfg.high_tide_push * (0.65 + rng() * 0.55);
					return true;
				case 'low-tide':
					this.timers['low-tide'] = jitterInt(rng, this.cfg.low_tide_dur, 0.3);
					this.timers['high-tide'] = 0;
					this.values.tide_bias = -this.cfg.low_tide_pull * (0.65 + rng() * 0.55);
					return true;
				case 'foam-burst':
					this.timers['foam-burst'] = jitterInt(rng, this.cfg.foam_burst_dur, 0.3);
					this.values.foam_gain = this.cfg.foam_burst_mult * (0.85 + rng() * 0.35);
					return true;
				case 'intro':
					this.timers['high-tide'] = 0;
					this.timers['low-tide'] = 0;
					this.timers['foam-burst'] = 0;
					this.timers.ending = 0;
					this.values.tide_bias = 0;
					this.values.foam_gain = 1;
					this.timers.intro = Math.max(1, Math.round(this.cfg.intro_dur));
					this.values.intro_total = this.timers.intro;
					return true;
				case 'ending':
					this.timers.intro = 0;
					this.timers['high-tide'] = 0;
					this.timers['low-tide'] = 0;
					this.timers['foam-burst'] = 0;
					this.values.tide_bias = 0;
					this.values.foam_gain = 1;
					this.timers.ending = Math.max(1, Math.round(this.cfg.ending_dur + Math.max(0, this.cfg.ending_linger)));
					this.values.ending_total = this.timers.ending;
					return true;
			}
			return false;
		}

		step() {
			this.tick++;
			for (const key of Object.keys(this.timers)) {
				if (this.timers[key] > 0) this.timers[key]--;
			}
			if ((!this.timers['high-tide'] || this.timers['high-tide'] <= 0) && (!this.timers['low-tide'] || this.timers['low-tide'] <= 0)) {
				this.values.tide_bias = 0;
			}
			if (!this.timers['foam-burst'] || this.timers['foam-burst'] <= 0) {
				this.values.foam_gain = 1;
			}
			api._helpers.paintProceduralGrid(this);
		}

		_tideLevel() {
			let level = 1;
			if (this.timers.intro > 0) {
				const total = this.values.intro_total || this.cfg.intro_dur;
				const progress = this._phaseProgress(total, this.timers.intro);
				level *= this.cfg.intro_tide + (1 - this.cfg.intro_tide) * progress;
			}
			if (this.timers.ending > 0) {
				const total = this.values.ending_total || (this.cfg.ending_dur + this.cfg.ending_linger);
				const progress = this._phaseProgress(total, this.timers.ending);
				level *= 1 - (1 - this.cfg.ending_wet) * progress;
			}
			return Math.max(0.05, level);
		}

		render(ctx, canvasW, canvasH, opts) {
			api._helpers.renderPixelGridEffect(this, ctx, canvasW, canvasH, opts);
		}
	}

	api.presets['beach'] = [
		{
			key: 'still-shore',
			label: 'still shore',
			config: {
				shoreline: 0.56,
				tide_amp: 3.2,
				wave_amp: 1.3,
				wave_freq: 0.14,
				speed: 0.05,
				slope: 0.08,
				foam: 0.24,
				shimmer: 0.18,
				hue: 196,
				hue_sp: 10,
				sat: 0.42,
				lmin: 0.26,
				lmax: 0.78,
			},
		},
		{
			key: 'gentle-tide',
			label: 'gentle tide',
			config: {
				shoreline: 0.58,
				tide_amp: 6,
				wave_amp: 2.4,
				wave_freq: 0.18,
				speed: 0.1,
				slope: 0.16,
				foam: 0.36,
				shimmer: 0.22,
				hue: 198,
				hue_sp: 16,
				sat: 0.5,
				lmin: 0.28,
				lmax: 0.82,
				high_tide_p: 0.0008,
				low_tide_p: 0.0006,
			},
		},
		{
			key: 'foamy-edge',
			label: 'foamy edge',
			config: {
				shoreline: 0.6,
				tide_amp: 7.4,
				wave_amp: 3.1,
				wave_freq: 0.21,
				speed: 0.12,
				slope: 0.2,
				foam: 0.5,
				shimmer: 0.18,
				hue: 194,
				hue_sp: 18,
				sat: 0.54,
				lmin: 0.3,
				lmax: 0.84,
				high_tide_p: 0.0012,
				foam_burst_p: 0.0013,
				foam_burst_mult: 2.2,
			},
		},
		{
			key: 'wide-beach',
			label: 'wide beach',
			config: {
				shoreline: 0.52,
				tide_amp: 4.8,
				wave_amp: 1.8,
				wave_freq: 0.12,
				speed: 0.08,
				slope: -0.1,
				foam: 0.3,
				shimmer: 0.28,
				hue: 202,
				hue_sp: 14,
				sat: 0.44,
				lmin: 0.24,
				lmax: 0.78,
				low_tide_p: 0.0011,
			},
		},
	];
	api.effects['beach'] = Beach;
})(window.AmbienceSim);

File: sim/beach.go
package sim

import (
	"fmt"
	"math"
	"sync"

	"github.com/romaine-life/ambience/rngutil"
)

type BeachConfig = ProceduralConfig

// BeachState is the shared procedural timer/value state plus the derived
// lifecycle contract field (computed at snapshot time, never restored).
type BeachState struct {
	ProceduralState
	Lifecycle Lifecycle `json:"lifecycle"`
}

// BeachSnapshot is the client-facing wire snapshot for Beach.
type BeachSnapshot struct {
	BeachState
	RNGState uint64 `json:"rngState,omitempty"`
}

// BeachPersistedState is the server-side resume state for Beach.
type BeachPersistedState struct {
	BeachState
	RNGState uint64 `json:"rngState"`
}

// Beach is a dedicated sim type for the beach effect, replacing the
// kind="beach" branch of the old Procedural type.
type Beach struct {
	mu sync.Mutex

	W, H int
	Grid [][]Pixel

	rng    *rngutil.RNG
	cfg    BeachConfig
	tick   int
	timers map[string]int
	values map[string]float64
	log    []LogEntry
}

var beachDefaultsLocal = BeachConfig{
	"intro_dur":       55,
	"intro_tide":      0.18,
	"ending_dur":      65,
	"ending_linger":   18,
	"ending_wet":      0.10,
	"shoreline":       0.58,
	"tide_amp":        6.0,
	"wave_amp":        2.4,
	"wave_freq":       0.18,
	"speed":           0.10,
	"slope":           0.16,
	"foam":            0.36,
	"shimmer":         0.22,
	"hue":             198,
	"hue_sp":          16,
	"sat":             0.50,
	"lmin":            0.28,
	"lmax":            0.82,
	"high_tide_p":     0.0,
	"low_tide_p":      0.0,
	"foam_burst_p":    0.0,
	"high_tide_dur":   60,
	"high_tide_push":  1.40,
	"low_tide_dur":    58,
	"low_tide_pull":   1.20,
	"foam_burst_dur":  34,
	"foam_burst_mult": 1.90,
}

func BeachSchema() EffectSchema {
	return EffectSchema{
		Name: "beach",
		Knobs: []Knob{
			{Key: "intro_dur", Label: "intro dur", Slot: SlotSpawn, Group: "introduction", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 55, Trigger: "intro",
				Description: "Ticks spent establishing the first advancing tide rhythm from a calmer shoreline."},
			{Key: "intro_tide", Label: "intro tide", Slot: SlotSpawn, Group: "introduction", Type: KnobFloat, Min: 0.05, Max: 0.5, Step: 0.01, Default: 0.18,
				Description: "Starting fraction of the full tide motion before the shoreline settles into rhythm."},
			{Key: "ending_dur", Label: "ending dur", Slot: SlotEnd, Group: "ending", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 65, Trigger: "ending",
				Description: "Ticks spent easing the waterline back out toward a calmer shore."},
			{Key: "ending_linger", Label: "ending linger", Slot: SlotEnd, Group: "ending", Type: KnobInt, Min: 0, Max: 140, Step: 5, Default: 18,
				Description: "Extra quiet ticks for wet sand and foam remnants to fade after the retreat."},
			{Key: "ending_wet", Label: "ending wet", Slot: SlotEnd, Group: "ending", Type: KnobFloat, Min: 0.02, Max: 0.35, Step: 0.01, Default: 0.10,
				Description: "Residual shoreline motion and wet-sand presence near the end of the outro."},
			{Key: "shoreline", Label: "shoreline", Slot: SlotLever, Group: "shore", Type: KnobFloat, Min: 0.42, Max: 0.76, Step: 0.01, Default: 0.58,
				Description: "Base shoreline height in the frame."},
			{Key: "tide_amp", Label: "tide amp", Slot: SlotLever, Group: "shore", Type: KnobFloat, Min: 1, Max: 14, Step: 0.5, Default: 6,
				Description: "How far the tide line advances and retreats."},
			{Key: "wave_amp", Label: "wave amp", Slot: SlotLever, Group: "shore", Type: KnobFloat, Min: 0.5, Max: 6, Step: 0.1, Default: 2.4,
				Description: "Height of the small shoreline ripples layered on top of the tide."},
			{Key: "wave_freq", Label: "wave freq", Slot: SlotLever, Group: "shore", Type: KnobFloat, Min: 0.05, Max: 0.35, Step: 0.01, Default: 0.18,
				Description: "Horizontal frequency of the shoreline wiggle."},
			{Key: "speed", Label: "tide speed", Slot: SlotLever, Group: "shore", Type: KnobFloat, Min: 0.02, Max: 0.3, Step: 0.01, Default: 0.10,
				Description: "How quickly the tide rhythm moves in and out."},
			{Key: "slope", Label: "shore slope", Slot: SlotLever, Group: "shore", Type: KnobFloat, Min: -0.35, Max: 0.35, Step: 0.01, Default: 0.16,
				Description: "Diagonal slant of the shoreline across the frame."},
			{Key: "foam", Label: "foam", Slot: SlotLever, Group: "shore", Type: KnobFloat, Min: 0.05, Max: 0.8, Step: 0.01, Default: 0.36,
				Description: "Brightness and thickness of the foam edge at the waterline."},
			{Key: "shimmer", Label: "shimmer", Slot: SlotLever, Group: "shore", Type: KnobFloat, Min: 0.02, Max: 0.6, Step: 0.01, Default: 0.22,
				Description: "Strength of the water-surface shimmer away from the foam line."},
			{Key: "hue", Label: "water hue", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 180, Max: 220, Step: 1, Default: 198,
				Description: "Base water hue. Lower values lean teal; higher values lean deep blue."},
			{Key: "hue_sp", Label: "hue spread", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0, Max: 28, Step: 1, Default: 16,
				Description: "Variation across water and foam accents."},
			{Key: "sat", Label: "saturation", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.1, Max: 0.9, Step: 0.01, Default: 0.50,
				Description: "Overall water color saturation."},
			{Key: "lmin", Label: "light min", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.1, Max: 0.6, Step: 0.01, Default: 0.28,
				Description: "Minimum lightness used for deeper water and wet sand."},
			{Key: "lmax", Label: "light max", Slot: SlotLever, Group: "color", Type: KnobFloat, Min: 0.4, Max: 0.95, Step: 0.01, Default: 0.82,
				Description: "Maximum lightness used for foam and bright shore reflections."},
			{Key: "high_tide_p", Label: "high tide", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "high-tide",
				Description: "Per-tick chance of the shoreline pushing farther inland for a while."},
			{Key: "low_tide_p", Label: "low tide", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "low-tide",
				Description: "Per-tick chance of the shoreline pulling farther back down the beach."},
			{Key: "foam_burst_p", Label: "foam burst", Slot: SlotEvent, Type: KnobFloat, Min: 0, Max: 0.02, Step: 0.0005, Default: 0, Trigger: "foam-burst",
				Description: "Per-tick chance of a brighter foamy wash crossing the edge."},
			{Key: "high_tide_dur", Label: "high dur", Slot: SlotEventMod, Group: "high-tide", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 60,
				Description: "Duration of the stronger high-tide push."},
			{Key: "high_tide_push", Label: "high push", Slot: SlotEventMod, Group: "high-tide", Type: KnobFloat, Min: 0.2, Max: 3, Step: 0.05, Default: 1.4,
				Description: "Additional inward shoreline offset during high tide."},
			{Key: "low_tide_dur", Label: "low dur", Slot: SlotEventMod, Group: "low-tide", Type: KnobInt, Min: 10, Max: 220, Step: 5, Default: 58,
				Description: "Duration of the lower-tide retreat."},
			{Key: "low_tide_pull", Label: "low pull", Slot: SlotEventMod, Group: "low-tide", Type: KnobFloat, Min: 0.2, Max: 3, Step: 0.05, Default: 1.2,
				Description: "Additional outward shoreline offset during low tide."},
			{Key: "foam_burst_dur", Label: "foam dur", Slot: SlotEventMod, Group: "foam-burst", Type: KnobInt, Min: 8, Max: 120, Step: 2, Default: 34,
				Description: "Duration of a brighter foamy edge."},
			{Key: "foam_burst_mult", Label: "foam x", Slot: SlotEventMod, Group: "foam-burst", Type: KnobFloat, Min: 1.05, Max: 3, Step: 0.05, Default: 1.9,
				Description: "Brightness multiplier applied during a foam burst."},
		},
	}
}

func defaultBeachConfig() BeachConfig { return cloneConfig(beachDefaultsLocal) }

func mergeBeachDefaults(cfg BeachConfig) BeachConfig {
	out := defaultBeachConfig()
	for k, v := range cfg {
		out[k] = v
	}
	if out["intro_dur"] <= 0 {
		out["intro_dur"] = beachDefaultsLocal["intro_dur"]
	}
	out["intro_tide"] = clamp01(out["intro_tide"])
	if out["ending_dur"] <= 0 {
		out["ending_dur"] = beachDefaultsLocal["ending_dur"]
	}
	if out["ending_linger"] < 0 {
		out["ending_linger"] = 0
	}
	out["ending_wet"] = clamp01(out["ending_wet"])
	if out["shoreline"] <= 0 {
		out["shoreline"] = beachDefaultsLocal["shoreline"]
	}
	if out["tide_amp"] <= 0 {
		out["tide_amp"] = beachDefaultsLocal["tide_amp"]
	}
	if out["wave_amp"] <= 0 {
		out["wave_amp"] = beachDefaultsLocal["wave_amp"]
	}
	if out["wave_freq"] <= 0 {
		out["wave_freq"] = beachDefaultsLocal["wave_freq"]
	}
	if out["speed"] <= 0 {
		out["speed"] = beachDefaultsLocal["speed"]
	}
	if out["foam"] <= 0 {
		out["foam"] = beachDefaultsLocal["foam"]
	}
	if out["shimmer"] <= 0 {
		out["shimmer"] = beachDefaultsLocal["shimmer"]
	}
	if out["hue"] == 0 {
		out["hue"] = beachDefaultsLocal["hue"]
	}
	if out["hue_sp"] < 0 {
		out["hue_sp"] = 0
	}
	if out["sat"] <= 0 {
		out["sat"] = beachDefaultsLocal["sat"]
	}
	if out["lmin"] <= 0 {
		out["lmin"] = beachDefaultsLocal["lmin"]
	}
	if out["lmax"] <= 0 {
		out["lmax"] = beachDefaultsLocal["lmax"]
	}
	if out["lmax"] < out["lmin"] {
		out["lmin"], out["lmax"] = out["lmax"], out["lmin"]
	}
	if out["high_tide_dur"] <= 0 {
		out["high_tide_dur"] = beachDefaultsLocal["high_tide_dur"]
	}
	if out["high_tide_push"] <= 0 {
		out["high_tide_push"] = beachDefaultsLocal["high_tide_push"]
	}
	if out["low_tide_dur"] <= 0 {
		out["low_tide_dur"] = beachDefaultsLocal["low_tide_dur"]
	}
	if out["low_tide_pull"] <= 0 {
		out["low_tide_pull"] = beachDefaultsLocal["low_tide_pull"]
	}
	if out["foam_burst_dur"] <= 0 {
		out["foam_burst_dur"] = beachDefaultsLocal["foam_burst_dur"]
	}
	if out["foam_burst_mult"] <= 0 {
		out["foam_burst_mult"] = beachDefaultsLocal["foam_burst_mult"]
	}
	return out
}

func NewBeach(w, h int, seed int64, cfg BeachConfig) *Beach {
	grid := make([][]Pixel, h)
	for i := range grid {
		grid[i] = make([]Pixel, w)
	}
	return &Beach{
		W:      w,
		H:      h,
		Grid:   grid,
		rng:    rngutil.New(seed),
		cfg:    mergeBeachDefaults(cfg),
		timers: make(map[string]int),
		values: make(map[string]float64),
	}
}

func (b *Beach) Resize(w, h int) {
	if w <= 0 || h <= 0 {
		return
	}
	b.mu.Lock()
	defer b.mu.Unlock()
	if w == b.W && h == b.H {
		return
	}
	b.W = w
	b.H = h
	b.Grid = make([][]Pixel, h)
	for i := range b.Grid {
		b.Grid[i] = make([]Pixel, w)
	}
}

func (b *Beach) SetConfig(cfg BeachConfig) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.cfg = mergeBeachDefaults(cfg)
}

func (b *Beach) EffectiveConfig() BeachConfig {
	b.mu.Lock()
	defer b.mu.Unlock()
	return cloneConfig(b.cfg)
}

func (b *Beach) Snapshot() BeachSnapshot {
	b.mu.Lock()
	defer b.mu.Unlock()
	return BeachSnapshot{
		BeachState: b.snapshotStateLocked(),
		RNGState:   b.rng.State(),
	}
}

func (b *Beach) RestoreSnapshot(snap BeachSnapshot) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.restoreStateLocked(snap.ProceduralState)
	if snap.RNGState != 0 {
		b.rng.SetState(snap.RNGState)
	}
}

func (b *Beach) SnapshotPersistedState() BeachPersistedState {
	b.mu.Lock()
	defer b.mu.Unlock()
	return BeachPersistedState{
		BeachState: b.snapshotStateLocked(),
		RNGState:   b.rng.State(),
	}
}

func (b *Beach) RestorePersistedState(ps BeachPersistedState) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.restoreStateLocked(ps.ProceduralState)
	if ps.RNGState != 0 {
		b.rng.SetState(ps.RNGState)
	}
}

func (b *Beach) snapshotStateLocked() BeachState {
	return BeachState{
		ProceduralState: ProceduralState{
			Tick:   b.tick,
			Timers: cloneTimerMap(b.timers),
			Values: cloneValueMap(b.values),
		},
		Lifecycle: b.lifecycleLocked(),
	}
}

// lifecycleLocked derives the effect-generic lifecycle contract value from
// the intro/ending timers. Beach's outro is non-terminal: when the ending
// timer expires the surf returns to steady state and automatic events
// resume, so lifecycle returns to running (the schema declares
// ending_terminal: false by omission).
func (b *Beach) lifecycleLocked() Lifecycle {
	switch {
	case b.timers["intro"] > 0:
		return LifecycleIntro
	case b.timers["ending"] > 0:
		return LifecycleEnding
	default:
		return LifecycleRunning
	}
}

func (b *Beach) restoreStateLocked(state ProceduralState) {
	b.tick = state.Tick
	b.timers = cloneTimerMap(state.Timers)
	if b.timers == nil {
		b.timers = make(map[string]int)
	}
	b.values = cloneValueMap(state.Values)
	if b.values == nil {
		b.values = make(map[string]float64)
	}
}

func (b *Beach) CurrentTick() int {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.tick
}

func (b *Beach) PerturbRNG(delta int64) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.rng.Mix(delta)
}

func (b *Beach) DrainLog() []LogEntry {
	b.mu.Lock()
	defer b.mu.Unlock()
	if len(b.log) == 0 {
		return nil
	}
	out := b.log
	b.log = nil
	return out
}

func (b *Beach) appendLog(kind, desc string) {
	b.log = append(b.log, LogEntry{Tick: b.tick, Type: kind, Desc: desc})
	if len(b.log) > 200 {
		b.log = b.log[len(b.log)-200:]
	}
}

func (b *Beach) intCfg(key string) int {
	return int(math.Round(b.cfg[key]))
}

func (b *Beach) TriggerEvent(name string) bool {
	b.mu.Lock()
	defer b.mu.Unlock()
	switch name {
	case "high-tide":
		b.startHighTideLocked("triggered")
	case "low-tide":
		b.startLowTideLocked("triggered")
	case "foam-burst":
		b.startFoamBurstLocked("triggered")
	case "intro":
		b.startIntroLocked()
		b.appendLog("intro", fmt.Sprintf("started (dur=%d, tide=%.2f)", b.timers["intro"], b.cfg["intro_tide"]))
	case "ending":
		b.startEndingLocked()
		b.appendLog("ending", fmt.Sprintf("started (fade=%d, linger=%d)", b.intCfg("ending_dur"), b.intCfg("ending_linger")))
	default:
		return false
	}
	return true
}

func (b *Beach) Step() {
	b.mu.Lock()
	defer b.mu.Unlock()

	b.tick++
	for key, value := range b.timers {
		if value > 0 {
			b.timers[key] = value - 1
		}
	}
	b.stepLocked()
}

func (b *Beach) startHighTideLocked(verb string) {
	b.timers["high-tide"] = jitterInt(b.rng, b.intCfg("high_tide_dur"), 0.3)
	b.timers["low-tide"] = 0
	b.values["tide_bias"] = b.cfg["high_tide_push"] * (0.65 + b.rng.Float64()*0.55)
	b.appendLog("high-tide", fmt.Sprintf("%s (dur=%d, bias=+%.2f)", verb, b.timers["high-tide"], b.values["tide_bias"]))
}

func (b *Beach) startLowTideLocked(verb string) {
	b.timers["low-tide"] = jitterInt(b.rng, b.intCfg("low_tide_dur"), 0.3)
	b.timers["high-tide"] = 0
	b.values["tide_bias"] = -b.cfg["low_tide_pull"] * (0.65 + b.rng.Float64()*0.55)
	b.appendLog("low-tide", fmt.Sprintf("%s (dur=%d, bias=%.2f)", verb, b.timers["low-tide"], b.values["tide_bias"]))
}

func (b *Beach) startFoamBurstLocked(verb string) {
	b.timers["foam-burst"] = jitterInt(b.rng, b.intCfg("foam_burst_dur"), 0.3)
	b.values["foam_gain"] = b.cfg["foam_burst_mult"] * (0.85 + b.rng.Float64()*0.35)
	b.appendLog("foam-burst", fmt.Sprintf("%s (dur=%d, x%.2f)", verb, b.timers["foam-burst"], b.values["foam_gain"]))
}

func (b *Beach) startIntroLocked() {
	b.timers["high-tide"] = 0
	b.timers["low-tide"] = 0
	b.timers["foam-burst"] = 0
	b.timers["ending"] = 0
	b.values["tide_bias"] = 0
	b.values["foam_gain"] = 1
	b.timers["intro"] = b.intCfg("intro_dur")
	b.values["intro_total"] = float64(b.timers["intro"])
}

func (b *Beach) startEndingLocked() {
	b.timers["intro"] = 0
	b.timers["high-tide"] = 0
	b.timers["low-tide"] = 0
	b.timers["foam-burst"] = 0
	b.values["tide_bias"] = 0
	b.values["foam_gain"] = 1
	endingTotal := b.intCfg("ending_dur") + max(0, b.intCfg("ending_linger"))
	if endingTotal < 1 {
		endingTotal = max(1, b.intCfg("ending_dur"))
	}
	b.timers["ending"] = endingTotal
	b.values["ending_total"] = float64(endingTotal)
}

func (b *Beach) stepLocked() {
	if b.timers["high-tide"] <= 0 && b.timers["low-tide"] <= 0 {
		b.values["tide_bias"] = 0
	}
	if b.timers["foam-burst"] <= 0 {
		b.values["foam_gain"] = 1
	}
	if b.timers["intro"] <= 0 {
		delete(b.values, "intro_total")
	}
	if b.timers["ending"] <= 0 {
		delete(b.values, "ending_total")
	}
	if b.timers["intro"] > 0 || b.timers["ending"] > 0 {
		return
	}
	if b.timers["high-tide"] <= 0 && b.timers["low-tide"] <= 0 && b.cfg["high_tide_p"] > 0 && b.rng.Float64() < b.cfg["high_tide_p"] {
		b.startHighTideLocked("started")
	}
	if b.timers["low-tide"] <= 0 && b.timers["high-tide"] <= 0 && b.cfg["low_tide_p"] > 0 && b.rng.Float64() < b.cfg["low_tide_p"] {
		b.startLowTideLocked("started")
	}
	if b.timers["foam-burst"] <= 0 && b.cfg["foam_burst_p"] > 0 && b.rng.Float64() < b.cfg["foam_burst_p"] {
		b.startFoamBurstLocked("started")
	}
}
